text/3541-async-drop.md

```diff
@@ -0,0 +1,225 @@
+
+#[tokio::main]
+async fn main() {
+    {
+        let _async_res = AsyncResource {
+            data: String::from("Example Data"),
+        };
+        // drop will be triggered here
+    }
+    // AsyncResource is dropped as it goes out of scope, and its asynchronous drop logic may be executed later by the Tokio runtime
+
+    // Sleep to ensure `drop` completed its execution
+    sleep(Duration::from_secs(2)).await;
+    println!("Main function completed");
+}
+```
+
+Moreover, to seamlessly integrate native async drop support into Rust, it is essential to outline how executors should handle asynchronous resource cleanup. This section also provides guidelines for developers using executors to ensure consistent and correct execution of async drop logic.
+
+1. **Recognition of AsyncDrop Trait**
+   Futures, such as those utilized in async Rust programming, should be updated to recognize types implementing the `AsyncDrop` trait. This can be achieved through a mechanism such as trait bounds or associated types to identify futures that require special handling for asynchronous cleanup.
+
+   ```rust
+   trait AsyncDrop {
+       async fn drop(&mut self);
+   }
+
+   impl AsyncDrop for MyFuture {
+       // Your asynchronous cleanup logic goes here
+       async fn drop(&mut self) {
+           println!("Async cleanup executed");
+       }
+   }
+   ```
+
+   Here, the `AsyncDrop` trait declares an asynchronous `drop` method, which should be implemented internally, allowing futures to define their asynchronous cleanup logic independently. In the example implementation for `MyFuture`, the `AsyncDrop` trait is implemented, and the asynchronous cleanup logic is executed when the `drop` method is called. This approach provides flexibility in handling resource management during asynchronous execution for different types of futures.
+
+1. **Async Drop Invocation**
+   Executors must ensure that the async `drop` method is invoked when a future implementing the `AsyncDrop` trait goes out of scope or completes. This involves tracking the lifecycle of futures and, upon termination, triggering the async drop logic before releasing associated resources.
+
+   ```rust
+   async fn execute_future<T: AsyncDrop + Future + Copy>(&self, mut fut: T) {
+       // Execute the future
+       let result = fut.await;
+
+       // Perform async drop before releasing resources
+       fut.drop().await;
+   }
+   ```
+
+   The `fut.await` call will execute the future (`fut`), pausing the current thread until the future completes and produces a result. Following the completion of the future's execution, the executor invokes the `drop` method on the future (`fut`). This ensures the execution of asynchronous drop logic, allowing the future to perform any necessary cleanup operations before resources are released. The `Copy` trait is used here to allow the future (`fut`) to be moved into the await call without consuming its ownership, ensuring that it can still be used later.
```

Comment:
`Copy` here just makes you drop a different copy of the `Future`. It also makes no sense in general.

This function also makes no sense. Why would you write such a function? Just do `fut.await`. Why do you need to manually drop `fut`? This removes the entire point of an `AsyncDrop`.

Reply:
Appreciate your reviews here. Agree with you on the `Copy` trait, it should be removed, my bad. But what I am trying to convey here is what the new `AsyncDrop` trait would accomplish internally: It should be called at the end before releasing the resource.

Again, thanks for your time and energy to criticize this RFC. I will give it a complete rewrite and address all the points raised here!